### PR TITLE
Fix: Contribute lambda times out fetching upstream subgraphs (#3613)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
https://github.com/DataBiosphere/azul/issues/3613

### Author

- [x] PR title references issue
- [ ] ~~PR title matches issue title (preceded by `Fix: ` for bugs)~~   <sub>the `Fix: ` prefix is appropriate here even though the issue is not labeled 'bug'</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Gathered performance data on the dcp11 snapshot as described [here](https://github.com/DataBiosphere/azul/issues/3613#issuecomment-961301758)
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [ ] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
> https://github.com/DataBiosphere/azul/issues/3613#issuecomment-961301758

On my machine, it took just under 9 minutes to can all 5 bundles. `fe70d276-05be-5ad7-8d87-135c04cfaf54` took the longest at 2:49.35.

**Review by achave11 — APPROVED** [review]
Approved, LGTM!

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Maybe I'm blind but I'm not seeing the batching mentioned in the ticket. 

Also

```diff
Index: src/azul/plugins/repository/tdr/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr/__init__.py b/src/azul/plugins/repository/tdr/__init__.py
--- a/src/azul/plugins/repository/tdr/__init__.py	(revision ba6f80f44c17da35ecf616bc6aeaa12c29b05686)
+++ b/src/azul/plugins/repository/tdr/__init__.py	(date 1636421357738)
@@ -5,6 +5,7 @@
     ThreadPoolExecutor,
 )
 import datetime
+from functools import partial
 from itertools import (
     groupby,
 )
@@ -415,11 +416,14 @@
                            entity_ids: Union[Set[EntityID], Set[BundleFQID]],
                            ) -> List[BigQueryRow]:
         """
-        Efficiently retrieve multiple entities from BigQuery in a single query
+        Efficiently retrieve multiple entities from BigQuery in a single query.
+
         :param source: Snapshot containing the entity table
+
         :param entity_type: The type of entity, corresponding to the table name
+
         :param entity_ids: For links, the fully qualified UUID and version of
-        each `links` entity. For other entities, just the UUIDs.
+                           each `links` entity. For other entities, just the UUIDs.
         """
         pk_column = entity_type + '_id'
         version_column = 'version'
@@ -434,21 +438,25 @@
         if entity_type == 'links':
             assert issubclass(entity_id_type, BundleFQID), entity_id_type
             entity_ids = cast(Set[BundleFQID], entity_ids)
-            search_columns = (pk_column, version_column)
-            search_values = (
-                (f'"{fqid.uuid}"', f'TIMESTAMP("{fqid.version}")')
+            where_columns = (pk_column, version_column)
+            # REVIEW: Use this where applicable, the nested string literals are giving me a headache
+            #         For SQL strings literals, prefer single quotes as they are more common and work better in JSON.
+            def quote(s):
+                return f"'{s}'"
+            where_values = (
+                (quote(fqid.uuid), f'TIMESTAMP({quote(fqid.version)})')
                 for fqid in entity_ids
             )
             expected = {fqid.uuid for fqid in entity_ids}
         else:
             assert issubclass(entity_id_type, str), (entity_type, entity_id_type)
-            search_columns = (pk_column,)
-            search_values = ((f'"{entity_id}"',) for entity_id in entity_ids)
+            where_columns = (pk_column,)
+            where_values = ((f'"{entity_id}"',) for entity_id in entity_ids)
             expected = entity_ids
         query = f'''
             SELECT {', '.join({pk_column, *non_pk_columns})}
             FROM {table_name}
-            WHERE {self._in(search_columns, search_values)}
+            WHERE {self._in(where_columns, where_values)}
         '''
         log.debug('Retrieving %i entities of type %r ...', len(entity_ids), entity_type)
         rows = self._query_latest_version(source, query, group_by=pk_column)
@@ -467,9 +475,9 @@
         >>> plugin._in(('foo', 'bar'), [('"abc"', '123'), ('"def"', '456')])
         '(foo, bar) IN (("abc", 123), ("def", 456))'
         """
-        columns = ', '.join(columns)
-        values = ', '.join(f'({", ".join(value)})' for value in values)
-        return f'({columns}) IN ({values})'
+        def join(i):
+            return '(' + ', '.join(i) + ')'
+        return join(columns) + ' IN ' + join(map(join, values))
 
     def _find_upstream_bundles(self,
                                source: TDRSourceRef,
Index: test/indexer/test_tdr.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/test_tdr.py b/test/indexer/test_tdr.py
--- a/test/indexer/test_tdr.py	(revision ba6f80f44c17da35ecf616bc6aeaa12c29b05686)
+++ b/test/indexer/test_tdr.py	(date 1636422173967)
@@ -290,14 +290,15 @@
         >>> plugin._in(('foo', 'bar'), [('"abc"', '123'), ('"def"', '456')])
         '(foo = "abc" AND bar = 123) OR (foo = "def" AND bar = 456)'
         """
+        # REVIEW: the triple quoted string had two literal characters in it. Not
+        #         worth the f'''''' expecially considering that the whitespace
+        #         around + also makes things easier to read.
         return ' OR '.join(
-            f'''({(
-                ' AND '.join(
-                    f'{column} = {value}'
-                    for column, value in zip(columns, value_slice)
-                )
-            )})'''
-            for value_slice in values
+            '(' + ' AND '.join(
+                f'{column} = {inner_value}'
+                for column, inner_value in zip(columns, value)
+            ) + ')'
+            for value in values
         )
 
 

```

**Comment by noah-aviel-dove** [comment]
The only limits that seem pertinent to the choice of batch size are the maximum allowed query size and maximum response size (https://cloud.google.com/bigquery/quotas#query_jobs). We aren't getting anywhere close to either of them. I cannot find any reference to a maximum size for BigQuery Arrays or the `IN` operator.

Maximum query size is 12MB. Our largest query is for retrieving 6,267 links (other queries retrieve more entities but this one grows faster due to referencing the versions in addition to the UUIDs). The query is only 527KB.

Maximum response size is 10 GB. It's difficult to know the exact size google calculates for our responses, and tedious to approximate, but I suspect files are the biggest culprit due to large number of entities combined with large content size per entity due to the descriptor fields (links are very small for the upstream bundles). I didn't check every query but retrieving 12,535 analysis files results results in a response size of only 11 MB when exported to CSV.

I did not obverse any performance improvement improvement when fetching the links for the largest bundle (6,267 stitched bundles) using a batch size of 10,000 (equivalent to no batching) vs 1,000 (both took about 2:45). Using a batch size of 100 was significantly slower (4:55).

All things considered, I'm not convinced that splitting the retrieval of the stitched bundles into batches of N actually provides any benefit.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 1]
```diff
@@ -354,83 +358,131 @@ def _stitch_bundles(self,
...
-            FROM {backtick(self._full_table_name(source, 'links'))}
-            WHERE links_id = '{links_id.uuid}'
-                AND version = TIMESTAMP('{links_id.version}')
-        '''))
-        links = dict(links)  # Enable item assignment to pre-parse content JSON
-        links['content'] = json.loads(links['content'])
+        source = one({fqid.source.spec for fqid in links_ids})
+        links = self._retrieve_entities(source, 'links', links_ids)
+        links = {
+            # Enable item assignment to pre-parse content JSON
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:405]:
```suggestion
            # Copy the values so we can reassign `content` below
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
> All things considered, I'm not convinced that splitting the retrieval of the stitched bundles into batches of N actually provides any benefit.

In this case it provides the benefit of not crashing when there are too many upstream bundles. Hardly anything scales without limits, so batching should be your default approach in any situation. It's just that sometimes we forget to consider scaling limits on a particular type of client-server exchange or disregard them because we know that other limits kick in first.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 2]
```diff
@@ -354,83 +358,134 @@ def _stitch_bundles(self,
...
-            FROM {backtick(self._full_table_name(source, 'links'))}
-            WHERE links_id = '{links_id.uuid}'
-                AND version = TIMESTAMP('{links_id.version}')
-        '''))
-        links = dict(links)  # Enable item assignment to pre-parse content JSON
-        links['content'] = json.loads(links['content'])
+        source = one({fqid.source.spec for fqid in links_ids})
+        links = self._retrieve_entities(source, 'links', links_ids)
+        links = {
+            # Copy the values for we can reassign `content` below
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:408]:
```suggestion
            # Copy the values so we can reassign `content` below
```

**Review by hannes-ucsc — APPROVED** [review]
@amarjandu can you make this one change before you push to sandbox?

File: src/azul/plugins/repository/tdr/__init__.py (hunk 2 above)
Comment (amarjandu) [reply src/azul/plugins/repository/tdr/__init__.py]: fixed typo here 9022abb

**Comment by amarjandu** [comment]
Reindex as part of https://gitlab.azul.data.humancellatlas.org/ucsc/azul/-/jobs/11849

4 of the 5 bundles that were supposed to be able to be reindexed were not reindexed due to memory constraints. See #3659

**Pull request merged**